components/st-microelectronics/generate-ioc.stanza

```diff
@@ -0,0 +1,351 @@
+
+public pcb-enum ocdb/st-microelectronics/generate-ioc/STM32IPs :
+  I2C1IP   ; I2C1
+  I2C2IP   ; I2C2
+  CAN1IP   ; CAN1
+  CAN2IP   ; CAN2
+  USART1IP ; USART1
+  USART2IP ; USART2
+  USART3IP ; USART3
+  UART4IP  ; UART4
+  UART5IP  ; UART5
+  USART6IP ; USART6
+  RCCIP    ; RCC - default
+  RTCIP    ; RTC - when adding 32kHz xtal?
+  SPI1IP   ; SPI1
+  SPI2IP   ; SPI2
+  SPI3IP   ; SPI3
+  SYSIP    ; SYS - default
+  NVICIP   ; NVIC - default
+  ADC1IP   ; ADC1
+  ADC2IP   ; ADC2
+  ADC3IP   ; ADC3
+  DACIP    ; DAC
+
+defn ioc-file-header () -> String :
+  val header = "#MicroXplorer Configuration settings - do not modify\nFile.Version=6\nboard=custom\n"
+  to-string(header)
+
+defn valid-instance (i : JITXObject) :
+  match(mpn?(i): String) :
+    if prefix?(mpn?(i) as String, "STM32") :
+      true
+  else :
+    false
+
+; defn add-comma-to-params (exist-params)
+;   if not exist-params :
+;     [""]
+;   else :
+;     [","]
+
+defn stm32-ip-name (ip : STM32IPs) -> String :
+  switch(ip) :
+    I2C1IP   : "I2C1" ; I2C1
+    I2C2IP   : "I2C2" ; I2C2
+    CAN1IP   : "CAN1" ; CAN1
+    CAN2IP   : "CAN2" ; CAN2
+    USART1IP : "USART1" ; USART1
+    USART2IP : "USART2" ; USART2
+    USART3IP : "USART3" ; USART3
+    UART4IP  : "UART4" ; UART4
+    UART5IP  : "UART5" ; UART5
+    USART6IP : "USART6" ; USART6
+    RCCIP    : "RCC" ; RCC - default
+    RTCIP    : "RTC" ; RTC - when adding 32kHz xtal?
+    SPI1IP   : "SPI1" ; SPI1
+    SPI2IP   : "SPI2" ; SPI2
+    SPI3IP   : "SPI3" ; SPI3
+    SYSIP    : "SYS" ; SYS - default
+    NVICIP   : "NVIC" ; NVIC - default
+    ADC1IP   : "ADC1" ; ADC1
+    ADC2IP   : "ADC2" ; ADC2
+    ADC3IP   : "ADC3" ; ADC3
+    DACIP    : "DAC" ; DAC
+
+
+defn stm32-spi-type (mode : SPIType) -> String :
+  switch(mode) :
+    SPIFullDuplexController : "Full_Duplex_Master"
+    SPIFullDuplexPeripheral : "Full_Duplex_Slave"
+    SPIHalfDuplexController : "Simplex_Bidirectional_Master"
+    SPIHalfDuplexPeripheral : "Simplex_Bidirectional_Slave"
+    SPIRxController :         "RX_Only_Simplex_Unidirect_Master"
+    SPIRxPeripheral :         "RX_Only_Simplex_Unidirect_Slave"
+    SPITxController :         "TX_Only_Simplex_Unidirect_Master"
+    SPITxPeripheral :         "TX_Only_Simplex_Unidirect_Slave"
+
+defn stm32-usart-mode (mode : USARTMode) -> String :
+  switch(mode) :
+    USARTSynchronous :    "Synchronous"
+    USARTAsynchronous :   "Asynchronous"
+    USARTSingleWire :     "Half_duplex(single_wire_mode)"
+    USARTMultiprocessor : "Multiprocessor_communication"
+    USARTIRDA :           "IrDA"
+    USARTSmartcard :      "SmartCard"
+    USARTSmartcardWClk :  "SmartCard_With_Clock"
+    USARTLIN :            "LIN"
+
+defn stm32-gpio-mode (mode : GPIOModes) -> String :
+  switch(mode) :
+    GPIOOutput :   "GPIO_Output"
+    GPIOInput :    "GPIO_Input"
+    GPIOAnalog :   "GPIO_Analog"
+    GPIOEventOut : "EVENTOUT"
+    GPIOExtI :     "GPXTI\n# can't currently write EXTI pins correctly."
```

Comment:
What's the limitation?

Reply:
There is a number associated with the GPIOExt statement that I believe we do not know a-priori.